bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -330,54 +350,47 @@ private int readFromLogChannel(long entryLogId, BufferedReadChannel channel, Byt
      * A thread-local variable that wraps a mapping of log ids to bufferedchannels
      * These channels should be used only for reading. logChannel is the one
      * that is used for writes.
+     * We use this Guava cache to store the BufferedReadChannel.
+     * When the BufferedReadChannel is removed, the underlying fileChannel's refCnt decrease 1,
+     * temporally use 1h to relax replace after reading.
      */
-    private final ThreadLocal<Map<Long, BufferedReadChannel>> logid2Channel =
-            new ThreadLocal<Map<Long, BufferedReadChannel>>() {
+    private final ThreadLocal<Cache<Long, EntryLogBufferedReadChannel>> logid2ReadChannel =
+            new ThreadLocal<Cache<Long, EntryLogBufferedReadChannel>>() {
         @Override
-        public Map<Long, BufferedReadChannel> initialValue() {
+        public Cache<Long, EntryLogBufferedReadChannel> initialValue() {
             // Since this is thread local there only one modifier
             // We dont really need the concurrency, but we need to use
             // the weak values. Therefore using the concurrency level of 1
-            return new MapMaker().concurrencyLevel(1)
-                .weakValues()
-                .makeMap();
+            return CacheBuilder.newBuilder().concurrencyLevel(1)
+                    .expireAfterAccess(readChannelCacheExpireTimeMs, TimeUnit.MILLISECONDS)
+                    //decrease the refCnt
+                    .removalListener(removal -> ((EntryLogBufferedReadChannel) removal.getValue()).release())
+                    .build();
         }
     };
 
-    /**
-     * Each thread local buffered read channel can share the same file handle because reads are not relative
-     * and don't cause a change in the channel's position. We use this map to store the file channels. Each
-     * file channel is mapped to a log id which represents an open log file.
-     */
-    private final ConcurrentMap<Long, FileChannel> logid2FileChannel = new ConcurrentHashMap<Long, FileChannel>();
+    @VisibleForTesting
+    long getReadChannelCacheExpireTimeMs() {
+        return readChannelCacheExpireTimeMs;
+    }
 
     /**
-     * Put the logId, bc pair in the map responsible for the current thread.
-     * @param logId
-     * @param bc
+     * Each thread local buffered read channel can share the same file handle because reads are not relative
+     * and don't cause a change in the channel's position.
+     * Each file channel is mapped to a log id which represents an open log file.
      */
-    public BufferedReadChannel putInReadChannels(long logId, BufferedReadChannel bc) {
-        Map<Long, BufferedReadChannel> threadMap = logid2Channel.get();
-        return threadMap.put(logId, bc);
+    private FileChannelBackingCache fileChannelBackingCache = new FileChannelBackingCache(this::findFile);
+    @VisibleForTesting
```

Comment:
Since you have a FakeEntryLogger in the test, why not just make fileChannelBackingCache package private, and use FakeEntryLogger from there?